### ISSUE TITLE
How to get rid of annoying popups in Windows 10?

### ISSUE THREAD

**Issue by Kitanotori** [Unitech/pm2#2182]
How do I disable the popup windows that pm2 generates in Windows 10 when starting, restarting or reloading a Node.js app? From the perspective of rapid server-side development, this is very annoying...

**Comment by Unitech** [comment]
Would it be an issue because of the way we spawn processes?

- https://github.com/Unitech/pm2/blob/development/lib/God/ClusterMode.js#L52
- https://github.com/Unitech/pm2/blob/development/lib/God/ForkMode.js#L94

I don't know if I have to customize something on this code segments,

**Comment by soyuka** [comment]
`{detached: true` should get rid of this error but it is not available for cluster.

**Comment by Unitech** [comment]
interesting, maybe @ThisIsMac47 will have a look at this on the next weeks

**Comment by vmarchaud** [comment]
After testing, the detached option doesnt hide popups in windows, i think that nodejs spawn a shell to spawn the process.
I will investigate more in the next days.

**Comment by vmarchaud** [comment]
Okay so : 
- In fork mode, you shouldnt have popups in the dev branch since now
- In cluster mode, unfortunately, we dont have control on the popup cause we use the [cluster implementation of nodejs](https://nodejs.org/dist/latest-v6.x/docs/api/cluster.html).

If you get any popup apart of cluster mode, respond to this issue to report it, i will investigate for each one.

**Comment by soyuka** [comment]
Because we can't get `detached: true` in cluster?

**Comment by vmarchaud** [comment]
They use the function fork, we can use ```detached: true``` only with spawn.

**Comment by soyuka** [comment]
Yes ok so if that's the main reason, the only solution is to port cluster into pm2 instead of using the nodejs cluster. 
Did you try a simple cluster implementation without pm2 to see if it pops up too?

**Comment by vmarchaud** [comment]
I wasnt able to reproduce with a simple cluster implementation, this might coming from the fact that we are a spawning a child from a forked child or something like this.
We manage to delete the popups in cluster mod by [rewriting the implementation](https://github.com/Unitech/pm2/commit/0f2ecdd4b5cfe305a0a58f5033a142c793fbb993) on the [branch of a forked cluster implementation](https://github.com/Unitech/pm2/tree/cluster-embedded), we will talk about how include this into future release of pm2.

**Comment by ajithr** [comment]
Is this issue fixed or any workaround available? I have faced the same issue when using cluster mode with watch option. The node window opens multiple time when saving the file.

**Comment by vmarchaud** [comment]
@ajithr No, like i said, the problem come from nodeJs implementation of the cluster mode. I suggest you to develop in fork mode.

**Comment by gustavohenke** [comment]
For me what has been working quite well is to use `--no-daemon`.
However, if pm2 was previously started as a daemon process, then this flag has no effect and the popups keep appearing, unless I run `pm2 kill`.

**Comment by centigrade-thomas-becker** [comment]
@gustavohenke I followed your instructions, but it did not work for me. I am using Windows 8.1 and pm2 2.0.19

**Comment by vmarchaud** [comment]
@centigrade-thomas-becker The fact a popup is spawning when you call `child_process.exec` isn't coming from pm2, it come from the system of spawning a child process of node itself, you may add `detached: true` to the spawning config and it will not spawn any popup.

The problem of #2182 is that in cluster mode, we can't control how the process are spawn, so we can't add this options to hide the popup under windows.

**Comment by soyuka** [comment]
I think I opened an issue in nodejs for this matter.

**Comment by soyuka** [comment]
@centigrade-thomas-becker you should try to use `child_process.spawn` and add a `{detached: true}` option.

**Comment by mswaminath** [comment]
I wanted to share what I did to overcome the popup.
Initially I will `npm start` my app that starts `pm2` as well, then I will do `pm2 stop all` immediately. After that when I again do`npm start`, it just pops up only once and never again.

**Comment by Kondax** [comment]
Do we have any updates on this for a proper implementation that will patch this issue?

I use PM2 when locally developing. It's becoming an issue to run `pm2 kill all` after every script run. Consoles interrupting me when working is not welcome.

**Comment by raghavmac** [comment]
Try this,
1) `npm start` your project as usual (pop-up's keep on coming up)
2) Stop the project (`Ctrl-C`)
3) `pm2 delete <project-name>`
4) Again `npm start` (this time it stops popping up)

Hope it works!

**Comment by Kondax** [comment]
I use the process json file to launch my pm2 instances so I do not believe
that will work for me.

**Comment by sant123** [comment]
Any fix to this? Just asking 😄

**Comment by centigrade-thomas-becker** [comment]
> I wanted to share what I did to overcome the popup.
> Initially I will npm start my app that starts pm2 as well, then I will do pm2 stop all immediately. After that when I again do npm start, it just pops up only once and never again.

@mswaminath Can you please post your pm2 config, and how you spawn your child processes?

**Comment by pvakharia9033** [comment]
Any permanent solution for this ??

**Comment by jgjake2** [comment]
This issue is especially terrible if you have autorestart enabled, and something causes your script to crash after it loads (I accidentally renamed a file the script frequently reads from). Then all hell breaks loose! Windows are spawning and crashing so fast, I couldn't type or click on anything since the windows spawn in the foreground with focus. It's completely ignoring "max_restarts" and "restart_delay" (i'm guessing these don't apply to clusters). I literally had no choice but to hard reset my computer.

**Comment by wayofthefuture** [comment]
Wow it feels like I just won a game of solitaire

**Comment by wayofthefuture** [comment]
So for production applications it is imperative to take the weight off the single event loop by forking a few processes to balance the load. Well, fork() does not take a detached option, and nor should it. Once the primary process fails the child processes will be closed subsequently, allowing pm2 to respawn the parent which will re-fork children, while keeping only 1 instance of each process and maintaining the IPC pipe between them. This is quite simple and the forking nature of the node parent process is beautiful in this regard in my opinion.

At this point our only option is two use pm2 to create 4 actual and separate node apps. This may work, maybe. I have no idea how we will accomplish any kind of IPC messages back and forth from the processes considering there is no parent process to route the messages.

In other words, pm2 for production purposes on windows is currently not working. I will mention that if anyone can find time to resolve this issue, we will be happy to finance your hourly work in the name of getting this problem solved ASAP.

Thank you.

**Comment by raghavmac** [comment]
I have found a very simple fix that am using now for development,

1. On your pm2 config file, just remove/comment the `instances: 1` line
2. `pm2 delete <project-name>`
3. `npm start`

Then you will see the magic, but I suggest to use this only for development.

**Comment by wayofthefuture** [comment]
We aren't using a config file. Our master process forks into sub-processes. Will this work for that?

**Comment by wayofthefuture** [comment]
SOLVED.

Use winsw and run "node myapp.js" as a windows service.

> [https://github.com/kohsuke/winsw](url)

**Comment by Kondax** [comment]
Any updates on this? I run multiple instances so cannot write off instance support.

**Comment by atrauzzi** [comment]
So, I don't know if you guys would consider it worth the extra dep, but I'm putting together a utility that needs to be able to spawn processes cross-platform.  I had an idea over the weekend to use edgejs!

![image](https://user-images.githubusercontent.com/115199/30088284-0aad71a4-926b-11e7-9064-344231dc7302.png)

Sure enough, it works without having to get an annoying console window for each thing launched.

**Comment by atrauzzi** [comment]
I've opened this ticket with nodejs: https://github.com/nodejs/node/issues/15217

Would be great if everyone watching this ticket went and upvoted and shared this one.  Really it seems like something that wouldn't be high impact for them to fix!

**Comment by gregorskii** [comment]
I think this is occurring for me on exec processes called by PM2 process management. I am using fork mode. What do you think, does this occur for subprocesses started with exec.

**Comment by milichev** [comment]
Any updates?
`child_process.spawn::windowsHide` option is already awailable:
https://github.com/nodejs/node/issues/15217#ref-pullrequest-267864790

**Comment by Unitech** [comment]
https://github.com/Unitech/pm2/pull/3255 https://github.com/nodejs/node/issues/15217#ref-pullrequest-267864790 thanks to @soyuka @milichev @cjihrig

**Comment by soyuka** [comment]
May you try development branch @milichev ?

**Comment by FireController1847** [comment]
FIX PLS

**Comment by FireController1847** [comment]
Is it possible to get this a little earlier than release?

**Comment by milichev** [comment]
Installed with no success. The same console fireworks -(

```
node -v
v8.9.0

yarn global add https://github.com/Unitech/pm2@development
```

**Comment by toddwong** [comment]
Any update?

**Comment by FireController1847** [comment]
It was fixed for me

**Comment by toddwong** [comment]
@FireController1847 Not for me I think.

```
git clone git@github.com:Unitech/pm2.git
cd pm2
npm link
cd \test
pm2 start test.js -i 2
```

Still got 2 black console windows poped up!

node --version
v9.2.0
pm2 --version
2.8.0

I'm on Window 10.

Is there any wrong with my setup?

**Comment by FireController1847** [comment]
Ah, actually, you're right. The window does popup. Nevermind.

**Comment by toddwong** [comment]
Unless i missed something,
God.nodeApp(in lib/God/ClusterMode.js) called cluster.fork, which has no windowsHide parameter at all!? Dead end!

**Comment by toddwong** [comment]
I got a hacky solution

```diff
diff --git a/lib/God.js b/lib/God.js
index bf3bedda..61178584 100644
--- a/lib/God.js
+++ b/lib/God.js
@@ -15,7 +15,7 @@
  ******************************/

 var semver        = require('semver');
-var cluster       = require('cluster');
+var cluster       = require('./cluster_hack');
 var numCPUs       = require('os').cpus() ? require('os').cpus().length : 1;
 var path          = require('path');
 var EventEmitter2 = require('eventemitter2').EventEmitter2;
diff --git a/lib/God/ClusterMode.js b/lib/God/ClusterMode.js
index 3a1185a3..af9bf43a 100644
--- a/lib/God/ClusterMode.js
+++ b/lib/God/ClusterMode.js
@@ -10,7 +10,7 @@
  * @author Alexandre Strzelewicz <as@unitech.io>
  * @project PM2
  */
-var cluster       = require('cluster');
+var cluster       = require('../cluster_hack');
 var cst           = require('../../constants.js');
 var Utility       = require('../Utility.js');
 var pkg           = require('../../package.json');
@@ -49,7 +49,7 @@ module.exports = function ClusterMode(God) {
       // { "args": ["foo", "bar"], "env": { "foo1": "bar1" }} will be parsed to
       // { "args": "foo, bar", "env": "[object Object]"}
       // So we passing a stringified JSON here.
-      clu = cluster.fork({pm2_env: JSON.stringify(env_copy), windowsHide: true});
+      clu = cluster.fork({pm2_env: JSON.stringify(env_copy), pm2_windowsHide: true});
     } catch(e) {
       God.logAndGenerateError(e);
       return cb(e);
diff --git a/lib/cluster_hack.js b/lib/cluster_hack.js
new file mode 100644
index 00000000..c1cfd86b
--- /dev/null
+++ b/lib/cluster_hack.js
@@ -0,0 +1,18 @@
+var child_process = require('child_process');
+var oldFork = child_process.fork;
+child_process.fork = function () {
+  var pos = 1;
+  if (pos < arguments.length && Array.isArray(arguments[pos]))
+    pos++;
+  if (pos < arguments.length && arguments[pos] != null
+          && typeof arguments[pos] === 'object'
+          && arguments[pos].env && arguments[pos].env.pm2_windowsHide) {
+    arguments[pos].windowsHide = true;
+    arguments[pos].silent = true;
+  }
+  return oldFork.apply(this, arguments);
+};
+var cluster = require('cluster');
+child_process.fork = oldFork;
+
+module.exports = cluster;

```

But still get black windows flashing when `pm2 stop x`  `pm2 kill`  `pm2 update` ...

**Comment by soyuka** [comment]
This should be fixed in nodejs then no?

**Comment by toddwong** [comment]
@soyuka  Not sure, but I've filed a feature request nodejs/node#17370 
But the black windows during `pm2 stop x` etc. should be address in pm2 I think.

**Comment by JoeTheFkingFrypan** [comment]
When using `fork` with `watch` activated, a black window flashes everytime a change is detected.
Is this normal behavior?

**Comment by sant123** [comment]
It seems is almost done https://github.com/nodejs/node/pull/17412

**Comment by dwelle** [comment]
Meanwhile, if this helps someone: `pm2@2.2.1` is the last version that doesn't show pop-ups for me on win7, `fork` mode (above that, it doesn't matter if I run `fork` mode or `cluster`, it always shows popups).

**Comment by Unitech** [comment]
Should work now, please try latest pm2 published:

```bash
$ npm install pm2 -g
$ pm2 update
```

**Comment by sant123** [comment]
Unfortunately continues spawning these windows 😞 

![image](https://user-images.githubusercontent.com/7959437/35253520-db04428c-ffb3-11e7-8c29-9e34cffc60a7.png)

Node: v9.4.0
OS: Windows 10
pm2: v2.9.2

Command: `pm2 start app.js -i max`

**Comment by sant123** [comment]
The server script:

``` js
const http = require("http");
const server = http.createServer(handler);
const port = 4875;

server.listen(port, () => {
    console.log("Listening at port %d", port);
});

function handler(req, res) {
    res.end("Holita desde Node.js");
}
```

**Comment by dwelle** [comment]
Still issues on `pm2@2.9.2` win7, config:

```
"exec_mode": "fork_mode",
"instances": 1
```

**Comment by metrotyranno** [comment]
Issue persists on windows 10 (pm2 version 2.9.3, node 9.4)

The prompts are only vissible for half a second though. Less than the 4 seconds of previous versions.

**Comment by wallet77** [comment]
PM2 3.0.0 has been released:

```
$ npm install pm2@latest -g
$ pm2 update
```

Be aware of breaking changes, especially drop support for node 0.12.
Full list here : https://github.com/Unitech/pm2/blob/master/CHANGELOG.md

**Comment by newlukai** [comment]
And why was this issue closed? After updating to pm2 3.0.0, I still have the same issue...

**Comment by wallet77** [comment]
It was closed cause we merged a PR which try to fix this issue.
We were waiting for feedback about it.

**Comment by soyuka** [comment]
> And why was this issue closed? After updating to pm2 3.0.0, I still have the same issue...

Mhh did you update pm2 in memory? the `windowsHide` feature added in nodejs should fix this.

**Comment by rastographics** [comment]
I updated to 3.0 by running ````npm install pm2 -g```` and then ````pm2 update```` . Window is still popping up for me.

**Comment by rastographics** [comment]
Should we be using a specific node version to fix this? I see that 8.8 added the windowsHide feature and I'm using 8.11.2.

**Comment by newlukai** [comment]
Still experiencing windows popping up with 3.0.3.
After starting a bunch of node services via `pm2 start <configFile>`, once in a while a cmd window pops up. Before I can get an idea what it displays, it's closed again. It seems this happens in a quite regular interval (every ~30 seconds).

**Comment by cleberar38** [comment]
Hey guys, 
After playing with this error for a little while, I fixed it by using the NODE.JS Child_Process Documentation.  

https://nodejs.org/api/child_process.html#child_process_child_process_exec_command_options_callback

In my app the first file to be read is the index.js on the root of the app. 

In the Documentation the lines look like this:
const { spawn } = require('child_process');

const subprocess = spawn(process.argv[0], ['child_program.js'], {
  detached: true,
  stdio: 'ignore'
});

subprocess.unref();
===========================================

But, 

I added the follow lines in my index.js file:

const { spawn } = require('child_process');

const subprocess = spawn(process.argv[0], ['index.js'], {
  stdio: 'ignore',
  windowsHide: true
});

subprocess.unref();
=========================================
Notice I remove the detached and add the windowsHide.
And it works like charm!
Good luck!
Cleber

**Comment by soyuka** [comment]
> It was closed cause we merged a PR which try to fix this issue.
We were waiting for feedback about it.

Shouldn't we document that when using a child process, the developper should use the `windowsHide` option?

**Comment by wallet77** [comment]
> Shouldn't we document that when using a child process, the developper should use the windowsHide option?

Problem is : we will copy/paste nodejs documentation. This work is not really ours but it's developer's responsibility. Anyway we can add a little explanation.

But I'm not sure everyone who has this issue is because of child process.

**Comment by newlukai** [comment]
With 3.1.2 it got even worse. Every 10 or 20 seconds, for each started app a window is (really fast) opened and closed. With 10 started apps, it's not really that usable.

**Comment by newlukai** [comment]
With 3.1.2 it got even worse. Every 10 or 20 seconds, for each started app a window is (really fast) opened and closed. With 10 started apps, it's not really that usable.

**Comment by inkhsutesou** [comment]
Are there any workarounds for this ?  I've tried using @cleberar38 's implementation but it still spawns popups.

**Comment by biels** [comment]
Same here with version 3.1.3. Cmd windows every 10 seconds. It leaves the computer unusable for any purpose other than serving.

**Comment by toddwong** [comment]
Only node >= 9.4.0 supports windowsHide in cluster.settings.  So this only works for node >= 9.4.0 with pm2 cluster mode. 
And you have to take care of you own child processes as well.

**Comment by liangwenzhong** [comment]
any solution ? cmd windows opened and closed.

**Comment by inkhsutesou** [comment]
@toddwong 

Hey, I apologize for asking for your time but would you be willing to write up instructions here on how to change this.

I am able to set PM2 to cluster, but I do not see where to place " windowsHide "
What does it mean to take care of your own child processes ?

Any answers are appreciated but there are enough users here that need a work around that it would help greatly to have this explained

Thank you !

**Comment by toddwong** [comment]
@inkhsutesou I mean if you call `child_process.fork/exec/...` in your own scripts, you'd have to set the `windowsHide` flag to true in the parameters. 
if you call `cluster.fork` in your own scripts, and you are working with node >= 9.4.0, you'd have to set the `windowsHide` flag to true when you call `cluster.setupMaster`. if you are working with node < 9.4.0, there isn't that `windowsHide` flag at all, and the black window may unavoidable.

**Comment by gk0r** [comment]
I am running standard, out of the box configuration on Windows 8 and Node 10.11. 
What do I need to do to make the console flashing stop?

**Comment by atuttle** [comment]
8.x is the current LTS branch. Was this fix merged into any specific 8.x release, or do we need to wait for 10.x LTS?

**Comment by duchu-net** [comment]
Latest version without this behavior for me is 3.0.3, every version to latest (3.2.2) - still occure.

**Comment by JoeTheFkingFrypan** [comment]
I can confirm what @duchu-net indicated
- no popup in `PM2 v3.0.3`
- Popups appear starting from `v3.1.0` all the way to `v3.2.2`

**Comment by atuttle** [comment]
@JoeTheFkingFrypan what version of Node?

**Comment by JoeTheFkingFrypan** [comment]
@atuttle I am using the following
- `Node v10.11.0`
- `Npm v6.4.1`
- `Yarn v1.9.4`

**Comment by atuttle** [comment]
Fair enough. Doesn't help those of us that would prefer to stay on Node LTS 8.x until 10.x becomes LTS though.

**Comment by TeoWay** [comment]
Guys I dont know how but after reinstaling PM2 i dont see this problem anymore. I also had problem with popups when launching gekko bot with pm2.

I cant say what i did, - Im only start study Node.js and else. To understand more how npm work i was delete pm2, install it local and global couple times. I was also deleted some catalogs and files in my user directory and in catalogs in AppData directory. 

Sorry for this useless post I just want to say that solution exist. And maybe some one with more experience will find the answer.

pm2 3.2.2
Windows 10
node v8.11.3
npm 5.6.0

**Comment by TeoWay** [comment]
Guys I dont know how but after reinstaling PM2 i dont see this problem anymore. I also had problem with popups when launching gekko bot with pm2.

I cant say what i did, - Im only start study Node.js and else. To understand more how npm work i was delete pm2, install it local and global couple times. I was also deleted some catalogs and files in my user directory and in catalogs in AppData directory. 

Sorry for this useless post I just want to say that solution exist. And maybe some one with more experience will find the answer.

pm2 3.2.2
Windows 10
node v8.11.3
npm 5.6.0

**Comment by majimboo** [comment]
Experienced this with a node and pm2 installation done today. Everything is the latest. Still too many annoying pop ups.

**Comment by yeikel** [comment]
I am also experiencing this issue. 

Node : v8.11.1
Pm2 : 3.2.2
OS : Windows 10

**Comment by hojakejake** [comment]
> I can confirm what @duchu-net indicated in [his comment](https://github.com/Unitech/pm2/issues/2182#issuecomment-427419862)
> 
> * no popup in [`PM2 v3.0.3`](https://github.com/Unitech/pm2/blob/master/CHANGELOG.md#303--tue-aug-07-2018-233505-gmt0200-cest-)
> * Popups appear starting from [`v3.1.0`](https://github.com/Unitech/pm2/blob/master/CHANGELOG.md#310--mon-sep-10-2018-102513-gmt0200-cest-) all the way to [`v3.2.2`](https://github.com/Unitech/pm2/blob/master/CHANGELOG.md#321-51018)
> 
> _Edited for clarity_: Running in the following environment
> 
> * `Node v10.11.0`
> * `Npm v6.4.1`
> * `Yarn v1.9.4`

it really works! thanks!

**Comment by matteoPhre** [comment]
I was experiencing this issue, and I solved with a json configuration to run my apps (I suggest to use this only in development mode)
`
{
  "apps": [
    {
      "name": "apigateway",
      "script": "./../apigateway/ApiGateway.js",
	  "exec_mode": "fork",
	  "env": {
		  "NODE_ENV":"development"
	  }
    },
    {
      "name": "mainSrv",
      "script": "./../main-srv/ExpressServer.js",
      "exec_mode": "fork",
	  "env": {
		  "NODE_ENV":"development"
	  }
    }
  ]
}

`
Hope it helps!

**Comment by msm1089** [comment]
I also get this issue. Thanks for peoples suggestions above to resolve it, but none of them worked for me. I used an ecosystem.config.js and saw identical behaviour to running the app without it.

Node: 10.10.0
PM2: 3.03, 3.2.2 (note: running in fork mode)
O/S: Windows 10
NPM: 6.4.1

@toddwong @cleberar38 
The windowsHide flag that was mentioned - how would I use that in my setup please? I am using pm2 to run a script from package.json currently. Can I still somehow edit the index.js to add that flag?

package.json
```
...
"scripts": {
    "build": "rimraf dist && babel src -d dist",
    "test:e2e": "dotenv -e envs/test.env -e envs/.env ./scripts/e2e.test.sh",
    "serve": "yarn run build && node dist/index.js"
  }
...
```

e2e.test.sh
```
#!/usr/bin/env bash
...
if ! netstat -aon | grep "0.0.0.0:$SERVER_PORT" | grep "LISTENING"; then
  pm2 -s start ./scripts/serve.js
  until netstat -aon | grep "0.0.0.0:$SERVER_PORT" | grep "LISTENING"; do
    sleep $RETRY_INTERVAL
  done
...
pm2 delete all
```

If anyone has other suggestions or thinks they can fix, I will happily test them out :)

**Comment by toddwong** [comment]
@msm1089 try this:
```
var exec = require('child_process').exec;
exec('npm run serve', {windowsHide: true});
```

**Comment by msm1089** [comment]
@toddwong 
Perfecto! I briefly see a couple of cmd windows open, but they go away and stay away :)

**Comment by navinmistry** [comment]
thanks for  figuring out this, 
I'm using node app with pm2 on windows server 2016 and 
both exec and spawn call are now not flashing any cmd popups. ( they are staying hidden ) 

pm2 version : 3.2.2
node 10.14.1

**Comment by BenjaminDish** [comment]
As @TheOptimisticFactory suggested, downgrading pm2 to 3.0.3 does the job for Node 8.x users.

Thank you !

**Comment by dekdevy** [comment]
reverting to PM2 3.0.3 indeed fixes this problem. Still worth a bump I think.

**Comment by newlukai** [comment]
Just tried the current version 3.3.1 and there it's getting worse. Pm2 doesn't stop to open and instantly close pop ups. It was quite difficult to enter "pm2 stop all".

**Comment by ColinTree** [comment]
Still annoyed by this issue..

**Comment by florian-kittel** [comment]
Last version where this not appears is "pm2": "3.0.3". That means, it was fixed and after 3.0.3 it is back again.

So I can't upgrade to 3.4.0. It is not possible to develop when the window pops up and closes on any change.

**Comment by snowwolfjay** [comment]
same issue.
v3.4.0 popup 4 window every second and disappear quickly too

**Comment by clarkttfu** [comment]
v3.0.3 does not work for me. But v2.10.4 (with node 6.4 ) does not pop at all.

**Comment by Unitech** [comment]
If anyone have a clue of which option should be passed to spawn function to avoid this behavior that is welcome.

Here is the culprit: https://github.com/Unitech/pm2/blob/master/lib/God/ForkMode.js#L98

**Comment by lllnnnggg555** [comment]
[child_process - windowsHide not working with detached: true](https://github.com/Unitech/pm2/issues/2182)
It seems a bug of node. Maybe you should change a way to start Daemon @Unitech. I use a detached process to start Daemon.js.Then run pm2 start, it do well.

**Comment by zveljkovic** [comment]
If you use `npx pm2 kill` and `npx pm2 start --no-daemon` to start the dev server editing
node_modules/pm2/lib/God/ForkMode.js:93 from `detached : false,` to `detached : true,` it doesn't create annoying popups. 

System: 
node -v v11.14.0
npm -v 6.7.0
pm2: "^3.5.0",
windows 10

**Comment by Cheney925** [comment]
Same issue. I've tried different node versions and different pm2 versions, it still doesn't work.

Node: 8.2.1, 10.12.0, 10.14.1, 10.16.0
PM2: 3.0.3, 3.5.1
Windows 10

If I run some simple code like this blow, it won't popup windows.

```
pm2 start test.js
```

```
setInterval(() => {
	console.log('test...');
}, 50);
```
In the opposite way that running a big system would cause so many popups and super fast. Only I can do is rebooting.

**Comment by ekantor** [comment]
As a workaround you can use pm2 on a linux subsystem. It works pretty well.

**Comment by htchtc052** [comment]
Any updates? I want develope node apps on windows

**Comment by pq1949** [comment]
try this environment
node 8.11.2
pm2 3.5.1

**Comment by Cheney925** [comment]
Hi, guys. I found a solution.
On Windows, you can try [pm2-windows-service](https://github.com/jon-hall/pm2-windows-service). When pm2 register as a Windows service, no popups anymore.

node 10.16.0
pm2 3.5.1
Windows 10

**Comment by martin-dimitrov13** [comment]
wtf

**Comment by bemontibeller** [comment]
No updates on this?

**Comment by florian-kittel** [comment]
No updates for newer version... from time to time I try it out when I update the other dependencies, but pm2 still give me the pop up windows over version 3.0.x. So I stay with this version.

Because of some users said, that they have the window popup also with 3.0.x here is my local setup:

In package.json
```
"scripts": {
  ...
  "pm2": "pm2",
  "start": "pm2 start ./process.json",
  ...
},
"dependencies": {
  ...
  "pm2": "3.0.x",
  ...
}
```

In process.json
```
{
  "apps": [
    {
      "name": "myMicroservice",
      "script": "./applications/myMicroservice/index.js",
      "watch": ["./applications/myMicroservice", "./libraries"],
      "merge_logs": true,
      "restart_delay": 1000,
      "log_date_format": "YYYY-MM-DD HH:mm Z"
    },
   ... // other processes with similar setup
} 
```

With this setup I havn't the windows pop up. I use the same setup on my linux server and only change environement variables. (Of course you can use a different setup for your server on your hoster.)

This works for me for development on Windows 10 and Windows 7.

**Comment by MohammedMars** [comment]
Any solution for this ??

**Comment by martyaxel** [comment]
A solution to this, would be very nice, or a workaround witouth having to install  vmware :|

**Comment by theCuriousOne** [comment]
> 
> 
> @msm1089 try this:
> 
> ```
> var exec = require('child_process').exec;
> exec('npm run serve', {windowsHide: true});
> ```
@MohammedMars   https://github.com/Unitech/pm2/issues/2182#issuecomment-441269031

**Comment by martyaxel** [comment]
I tried it, it doesnt work. I managed to make a workaround. I installed WSL (windows subsystem linux, ubuntu) on windows, and used the linux console to run my programs, it automatically connected to xampp and so on, i use it just like i used cmd on windows with a lot more features.

**Comment by theCuriousOne** [comment]
@martyaxel I have successfully managed to get it to work with the above code (lets say that is a startup1.js) which is called inside the ecosystem.config.js
```
module.exports = {
    apps: [
        {
            name: 'startup1',
            script: './startup1.js',
            exec_mode: 'fork',
            ....
         }
    ]
}
```

**Comment by youyinnn** [comment]
@theCuriousOne It is weird that I stop the popup cmd/powershell windows by setting the `exec_mode` to `'cluster'` and limit its as 1 instance:

``` javascript
module.exports = {
  apps : [{
    name: 'winmon',
    script: 'index.js',
    watch: false,
    instances  : 1,
    exec_mode: 'cluster'
  }],
};
```

**Comment by theCuriousOne** [comment]
@youyinnn Actually that one is even better. I just put  `script: 'npm run start'` and it works. Also it reads the stdout of Node.Js correctly. My previous example, did not did that.

**Comment by florian-kittel** [comment]
Hi guys,

I've installed pm2@4.4.x today and it seems, that it works without popups on windows now. Would be cool, if someone esle can confirm as well. I will need to test this on other PCs as well. 
Serverside I uses Linux, so there was no popups showing, but in lokal development it was annoying. That was the reason why I use 3.0.x since now, as I mentions a year ago.

Here is an example how my development config for pm2 that works on windows looks like
```
    {
      "name": "gateway",
      "script": "./applications/gateway/index.js",
      "watch": ["./applications/gateway", "./libraries"],
      "merge_logs": true,
      "restart_delay": 1000,
      "log_date_format": "YYYY-MM-DD HH:mm Z"
    },
```

**Comment by theCuriousOne** [comment]
@florian-kittel So you are using Windows OS for local development and you are seeing a terminal window popping up when you run pm2. right? how does your index.js look like?

**Comment by florian-kittel** [comment]
> 
> 
> @florian-kittel So you are using Windows OS for local development and you are seeing a terminal window popping up when you run pm2. right? how does your index.js look like?

I will do an abstract of the code, I testet it with that as well. It is the same that I uses a year ago when I mentioned that I need to use pm2@3.0.x to avoid the popups on windows, some comments above:

```
const express = require('express');

// removed futher initialisation code

const app = express();
const port = `${process.env.NODE_ENV_PORT_PREFIX || '30'}00`;

// removed code for routes

app.use('*', (req, res) => {
  res.status(404).send({
    error: 'Not Found',
    method: req.method,
    status: 404,
    system: 'gateway',
    url: req.baseUrl,
  });
});

app.listen(port, () => {
  console.log(`Gateway listening on port ${port}`);
});

```

I use node v12.15.0 and npm v6.14.5 on this machine.

**Comment by theCuriousOne** [comment]
Aha so you (@florian-kittel ) are calling the server directly from pm2. I have made an intermediate file that executes the `npm run command`

    const { spawn } = require('child_process');
    const exec = commands => {
        spawn(commands, ['run', 'start'], { stdio: 'inherit', shell: true, windowsHide: true });
    };
    exec('npm.cmd');

Disclaimer: I am also new to the pm2 system, and I am still looking for better solutions then the one above.

**Comment by RussellGrew** [comment]
I found this thread when looking for a way of starting PM2 following machine reboots without having to log in. As presumably the console popups require a user session in which to pop up.

I have been able to run scheduled tasks as SYSTEM to start PM2 without first logging in. This is with local node_modules PM2. And the schedule task starting in the project folder.

```
node node_modules/pm2/bin/pm2 start ...
```

Interestingly I cannot do the same to resurrect. Even with a PM2_HOME set.

I hope this helps someone.

**Comment by xnorpx** [comment]
Upgraded to 4.4.x on my windows server and windows still pop ups :(

**Comment by sreenivassign** [comment]
Hey any solution for this. I am using cluster mode.Any changes required to pm2/lib/God/ClusterMode.js  to resolve this or switching to linux subsystem helps

**Comment by Ivanca** [comment]
So, this problem only happens when you ask PM2 to run an app that itself spawns child processes, this issue seems to be related to the way Windows inherits visibility properties between processes so its not a problem with PM2 directly but mostly the way Windows works, in my case the solution was to find the `spawn` call creating the issue and adding `{windowsHide: true}` as third parameter. Good luck

**Comment by xinthose** [comment]
For some reason, I only get fast command prompt popups when the application is running from an administrator level directory (i.e. C:\Program Files (x86)).

**Comment by Unitech** [comment]
Hi,

Can someone try the development branch and let me know if the problem is fixed:

```
npm install Unitech/pm2#development -g
pm2 update
```

Then let me know if popups have disapeared.
Thank you

**Comment by Nekonyx** [comment]
@Unitech unfortunately it didn't help.

**Comment by Newtoxton** [comment]
> Any fix to this? Just asking 😄
pm2 kill is the answer to that, and to reproduce the issue, you run more than on pm2 thread. So they stack on top of one another. Solution is to kill them all and only have the intended one be started and run alone.

**Comment by BartaG512** [comment]
Any solution? I am still getting these annoying pop ups. 
I experience these popups to reproduce I start a simple file from programmatic api.

```
const pm2 = require('pm2');

pm2.connect((err) => {
	console.log('err', err);
	pm2.start('./test_script.js', { autorestart: false }, (err) => {
		console.log('start', err);
		pm2.disconnect((err) => {
			console.log('err', err);
		});
	});
});
```
The `test_script.js` contents is: `console.log('test')`

Versions:
pm2 4.5.6
node: v14.16.1
windows 10

**Comment by Notmeomg** [comment]
Ran into this issue, was able to resolve it by checking `pm2 logs`. In my case multiple servers were missing packages, after installing packages pm2 logs are clean and all is working.

**Comment by Cologler** [comment]
I wrote a [program](https://github.com/Cologler/noc-rust) to resolve this issue.

Just use `pm2 start noc -- app arg1 arg2 ...` to replace `pm2 start app -- arg1 arg2 ...`.

**Comment by jermitval** [comment]
```
npm i -g pm2@4.4
pm2 update
```
try it!

**Comment by Lemming8** [comment]
> @theCuriousOne It is weird that I stop the popup cmd/powershell windows by setting the `exec_mode` to `'cluster'` and limit its as 1 instance:
> 
> ```js
> module.exports = {
>   apps : [{
>     name: 'winmon',
>     script: 'index.js',
>     watch: false,
>     instances  : 1,
>     exec_mode: 'cluster'
>   }],
> };
> ```

this worked for me yesterday, but now it magically stopped working

**Comment by florian-kittel** [comment]
> > @theCuriousOne It is weird that I stop the popup cmd/powershell windows by setting the `exec_mode` to `'cluster'` and limit its as 1 instance:
> > ```js
> > module.exports = {
> >   apps : [{
> >     name: 'winmon',
> >     script: 'index.js',
> >     watch: false,
> >     instances  : 1,
> >     exec_mode: 'cluster'
> >   }],
> > };
> > ```
> 
> this worked for me yesterday, but now it magically stopped working

Didn't work on my side as well. 

Only version that do not pop up CMDs is 4.5.0

**Comment by theCuriousOne** [comment]
I really don't understand what are you trying to say @fireMan-34

- If you have the solution, then post it in this discussion
- If you found a problem in the code, then create a PR
- If you are trying something else (phishing, boasting, or what not), please stop and don't waste anyone's time

**Comment by jakejakeho** [comment]
The easiest solution will be switch to Linux :)

**Comment by Concara3443** [comment]
This problem has a solution this days?

**Comment by ronanorozco** [comment]
> ```
> npm i -g pm2@4.4
> pm2 update
> ```
> 
> try it!

Only thing that worked for me. Thanks

**Comment by clinton-mwachia** [comment]
Hello guys, any solution yet? I am using pm2 version 5.2.2 and node v16.18.0

**Comment by megaacheyounes** [comment]
still no solution, this issue was opened 6 years ago!

![fuck](https://www.reactiongifs.com/r/f-p.gif)

**Comment by carloslaborde** [comment]
Welp back to nodemon for development I guess

**Comment by GabrielLoures** [comment]
> 

Worked for me so far! Thanks

**Comment by Cheyenne55** [comment]
> ```
> npm i -g pm2@4.4
> pm2 update
> ```
> 
> try it!

Hi,

I have just tested this solution few minutes ago with node v18.12.1 and it works like a charm.
So the question is : why this particular version does not have the issue while the lastest still does have it ?
Looks like a regression to me.

**Comment by clinton-mwachia** [comment]
I also switched to Linux and it's working fine. I am using wsl2.

**Comment by githubUser982** [comment]
I suggest looking at for forever solved this problem, since forever does not not have it.

**Comment by jasonayerp** [comment]
> ```
> npm i -g pm2@4.4
> pm2 update
> ```
> 
> try it!

Confirmed this resolved the issue for me too.

**Comment by dr3adx** [comment]
piece of shit of a repo when u cant fix simple issue like this that EVERYONE is experiencing

**Comment by dr3adx** [comment]
fkin monkey developers

**Comment by richiedev666** [comment]
> `{detached: true` should get rid of this error but it is not available for cluster.

Perfect!

**Comment by Faf4a** [comment]
> ```
> npm i -g pm2@4.4
> pm2 update
> ```
> 
> try it!

Thanks! This worked just fine for me.. struggled with this issue for multiple months now.

**Comment by githubUser982** [comment]
Interesting, I wonder why it works for some.

**Comment by Tosinkoa** [comment]
Please any update on this?

**Comment by githubUser982** [comment]
There's no consensus. 4.4 works for some, not for others, [node-hide-console-window](https://www.npmjs.com/package/node-hide-console-window) is an option when it doesn't. I haven't tried noc.

I can only conclude that the maintainers don't care for windows, I suppose.

**Comment by tooolz** [comment]
> Welp back to nodemon for development I guess

this did it for me.
pm2 seems nice for a production environment, but for developing on windows? forget about it, nodemon to the rescue

**Comment by ve3** [comment]
I'm using pm2 v 5.3.0

It keep popup powershell window on every socket.io emit the message in node app. (popup and closed but very annoying because I can't focus on other window while it pops up.)

Is this related to this issue?

**Comment by hyroqi** [comment]
PSA, This window will keep popping up if your app has a crash or problem, I used nodemon to determine what was causing the crash (and the window popup), fixed it and pm2 stopped the unexpected behaviour

**Comment by hicodify** [comment]
I am using version 5.3.0 and I can confirm that the problem continues to occur. By any chance do you know if there is a solution for this on Windows?

**Comment by vite-vit** [comment]
You can run both source and pm2 in git bash. Not in cmd or powershell

**Comment by wxfred** [comment]
A python cmd window always popups in Windows 10 when I do the **pm2 start ecosystem.config.js**, how to avoid it

```
// ecosystem.config.js
module.exports = {
  apps: [
    {
      name: 'python',
      script: 'poetry',
      args: ['run', 'python', 'server.py'], // a python flask server
      instances: 1,
      exec_mode: 'fork',
      autorestart: true,
      watch: false,
    },
  ],
};
```

**Comment by DVLP** [comment]
Took me 20mins. Such a trivially fixable but very annoying issue shouldn't be hanging for so long

As @jermitval pointed out the last working version was 4.4
Between versions 4.4 and 4,5 a few dependencies were updated. Looks like the offending one is Vizion

Change Vizion to 0.12.3 and the problem disappears
Run
`where node`
i.e. C:\Program Files\nodejs\node.exe
open any terminal and go to `node_modules\pm2` subfolder i.e. `C:\Program Files\nodejs\node_modules\pm2`
`npm i vizion@0.2.13`
Done

**Comment by marui-obj** [comment]
In my case I have to flush terminal every interval time, so every time it flushes it keep popups annoying terminal at same interval time as it flush, my solutions is just removing that line from my source code.

This is my testcase in Python

```
# test.py
import time
from os import system, name

def flush():
    if name == "nt":
        _ = system("cls")
    else:
        _ = system("clear")

def main():
    print("foo")
    flush() #Comment me
    time.sleep(1)

if __name__ == "__main__" :
    while(True):
        main()
```

**Comment by matys1** [comment]
@DVLP setting `vizion: false` in the config file indeed solved the annoying popups in dev. Though that's really a workaround and not a proper fix. Bizzare that this issue has been open for 6 years.

**Comment by abaksha-sc** [comment]
@matys1 , thank you! Adding `vision: false` really solved the problem when parameter `--no-vizion` has no effect. Maybe because for development instance I use command `Debug` right from VSCode:

![image](https://github.com/user-attachments/assets/c5e77eee-a23a-44b8-822d-c15b64eccb38)

**Comment by Unitech** [comment]
I have disabled vision by default in the prochain release of PM2,
Could you make a try and let me know if that works for you:

```
npm install pm2-beta -g
pm2 update
```

Or with bun:

```
bun install pm2-beta -g
pm2 update
```

Thanks

**Comment by SrLiath** [comment]
I come here again to say: just run it with 'npx' and the screen won't show up.
